src/operation.js

```diff
@@ -300,21 +302,41 @@ export class Operation {
         attributes.homeDomain = opts.homeDomain;
 
         if (opts.signer) {
-            if (opts.signer.address) {
-                console.warn("signer.address is deprecated. Use signer.pubKey instead.");
-                opts.signer.pubKey = opts.signer.address;
+            let weight = this._checkUnsignedIntValue("signer.weight", opts.signer.weight, weightCheckFunction);
+            let key;
+
+            let setValues = 0;
+
+            if (opts.signer.pubKey) {
+                if (!StrKey.isValidPublicKey(opts.signer.pubKey)) {
+                  throw new Error("signer.pubKey is invalid.");
+                }
+                let rawKey = StrKey.decodePublicKey(opts.signer.pubKey);
+                key = new xdr.SignerKey.signerKeyTypeEd25519(rawKey);
+                setValues++;
             }
 
-            if (!Keypair.isValidPublicKey(opts.signer.pubKey)) {
-                throw new Error("signer.pubKey is invalid");
+            if (opts.signer.hashTx) {
+                if (!(Buffer.isBuffer(opts.signer.hashTx) && opts.signer.hashTx.length == 32)) {
+                    throw new Error("signer.hashTx must be 32 bytes Buffer.");
+                }
+                key = new xdr.SignerKey.signerKeyTypeHashTx(opts.signer.hashTx);
+                setValues++;
             }
 
-            opts.signer.weight = this._checkUnsignedIntValue("signer.weight", opts.signer.weight, weightCheckFunction);
+            if (opts.signer.hash) {
+                if (!(Buffer.isBuffer(opts.signer.hash) && opts.signer.hash.length == 32)) {
+                    throw new Error("signer.hash must be 32 bytes Buffer.");
+                }
+                key = new xdr.SignerKey.signerKeyTypeHashX(opts.signer.hash);
+                setValues++;
+            }
 
-            attributes.signer = new xdr.Signer({
-                pubKey: Keypair.fromAccountId(opts.signer.pubKey).xdrAccountId(),
-                weight: opts.signer.weight
-            });
+            if (setValues != 1) {
+                throw new Error("Signer object must contain exactly one of signer.pubKey, signer.hash, signer.tx must be set.");
```

Comment:
Please improve grammar here:
Signer object must contain exactly one of signer.pubKey, signer.hash, signer.tx
or
Exactly one of signer.pubKey, signer.hash, signer.tx must be set.

It looks like you merged these two sentences